subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java

```diff
@@ -73,14 +74,18 @@ public ApiGroovyCompiler(Compiler<JavaCompileSpec> javaCompiler) {
     }
 
     private static abstract class IncrementalCompilationCustomizer extends CompilationCustomizer {
-        static IncrementalCompilationCustomizer fromSpec(GroovyJavaJointCompileSpec spec) {
-            if (spec.getCompilationMappingFile() != null) {
+        static IncrementalCompilationCustomizer fromSpec(GroovyJavaJointCompileSpec spec, File[] sortedSourceFiles) {
+            if (spec.getCompilationMappingFile() != null && !isJavaGroovyJointCompilation(sortedSourceFiles)) {
```

Comment:
I wonder if we could be more lenient here. It would be great if we only need to do a full recompilation if a Java source file changed.

What are the challenges with Java source files:
1. Groovy source file changed, its classes affect the Java source file: IIUC, we wouldn't be able to map the affected class file to a source file, so we would do a full recompilation
2. Java source file changed: We don't know which classes are affected, so we need to do a full recompilation

IIUC, the first case already is implemented. So what we need to do is to implement the second case: Whenever a Java (or more generally a non-Groovy) source file is affected, then the compilation will be a full recompilation.

Would this work? Is there a case I am missing?

Always doing a full recompilation when we have Java source files, then using incremental Groovy compilation on for our Spock tests won't be much fun anymore, since we mostly have some Java sources lying around...

Reply:
I can't think of a counterexample. Let's try to write some tests.

Reply:
@wolfs PTAL!